src/k8s-engine/__tests__/k3sHelper.spec.ts

```diff
@@ -265,4 +266,129 @@ describe(K3sHelper, () => {
+      });
+      test('can choose closest pre-release version when crossing from single digits into 10s', () => {
+        const semvers = convertToSemvers(['v1.2.3+k3s3', 'v1.2.3+k3s8', 'v1.2.3+k3s12']);
+
+        expect(subject['pickClosestVersion'](...semvers)).toBe(semvers[2]);
+      });
+    });
+
+    describe('selectClosestSemVer', () => {
+      // Note filenames are specified deliberately out of order in order to test sorting
+      test('finds the closest major version', () => {
+        const desiredSemver = new semver.SemVer('v3.1.2+k3s3');
+        const cachedFilenames = ['v1.2.9+k3s4', 'v4.2.8+k3s1', 'v4.3.0+k3s1', 'v1.2.9+k3s1'];
+        const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+        expect(selectedSemVer.raw).toBe(cachedFilenames[1]);
+      });
+      test('finds the closest minor version', () => {
+        const desiredSemver = new semver.SemVer('v1.12.2+k3s3');
+        const cachedFilenames = ['v1.23.9+k3s4', 'v2.12.8+k3s1', 'v1.7.0+k3s1', 'v1.2.9+k3s1'];
+        const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+        expect(selectedSemVer.raw).toBe(cachedFilenames[2]);
+      });
+      describe('finds the closest patch version', () => {
+        test('at the start of the list of existing names', () => {
+          const desiredSemver = new semver.SemVer('v1.12.2+k3s3');
+          const cachedFilenames = ['v1.12.9+k3s4', 'v1.12.8+k3s1', 'v1.12.4+k3s4', 'v1.12.4+k3s1'];
+          const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+          expect(selectedSemVer.raw).toBe(cachedFilenames[3]);
+        });
+        test('inside the list', () => {
+          const desiredSemver = new semver.SemVer('v1.12.10+k3s99');
+          const cachedFilenames = ['v1.12.9+k3s1', 'v1.12.8+k3s1', 'v1.12.4+k3s4', 'v1.12.4+k3s1'];
+          const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+          expect(selectedSemVer.raw).toBe(cachedFilenames[0]);
+        });
+        test('at the end of the list', () => {
+          const desiredSemver = new semver.SemVer('v1.12.9+k3s5');
+          const cachedFilenames = ['v1.12.9+k3s4', 'v1.12.8+k3s1', 'v1.12.4+k3s4', 'v1.12.4+k3s1'];
+          const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+          expect(selectedSemVer.raw).toBe(cachedFilenames[0]);
+        });
+      });
+      describe('finds the closest build version', () => {
+        test('over single digits', () => {
+          const desiredSemver = new semver.SemVer('v1.2.9+k3s3');
+          const cachedFilenames = ['v1.2.9+k3s4', 'v1.2.8+k3s1', 'v1.3.0+k3s1', 'v1.2.9+k3s1'];
+          const selectedSemVer = subject['selectClosestSemVer'](desiredSemver, cachedFilenames);
+
+          expect(selectedSemVer.raw).toBe(cachedFilenames[0]);
+        });
+        test('over double digits', () => {
```

Comment:
I don't understand why we're picking between build versions — we only store up to the patch version in the prefs, and always pick the highest build version.

Reply:
This was all based on what I observed:
- With networking on:
- Factory-reset clean everything
- npm run dev
- shutdown
- `k3s/v1.23.8+k3s1` is the only directory in the cache
- Disable network
- npm run dev

I got an exception that it couldn't download `v1.23.8+k3s2`

I can change things to always pick the highest existing build version ... that would simplify the code a bit.

Reply:
Now keeping only the highest build version for a run of same-major/minor/patch versions.